### DISCUSSION TITLE
Jetpack 5 and mender

### DISCUSSION THREAD

**Discussion by cmcquinn** [OE4T/tegra-demo-distro#231] [Q&A]
Commit [7d907b0](https://github.com/OE4T/tegra-demo-distro/commit/7d907b044d4bc110ab4d548a1015fbe6d90c1556) removed the tegrademo-mender disto stating that the existing mender integration will not work with Jetpack 5. What specifically is incompatible and is there some other way to integrate mender with Jetpack 5? I am working on a project currently based off of `dunfell-l4t-r32.5.0` and would like to move to the latest `kirkstone` branch to get Jetpack 5 but mender updates are a feature that we use frequently. I would be willing to help make this work if it is possible.

Thanks,
Cameron

**Comment by dwalkes** [comment]
See https://github.com/OE4T/meta-tegra/discussions/1006#discussioncomment-3729440 for the latest status I know about.  Help would be welcome if you'd like to take this on.

**Reply by cmcquinn** [reply]
What do I need to do to get started on this?

**Reply by dwalkes** [reply]
I'd suggest starting by reviewing the discussion in the monthly meeting and the conversations on gitter linked above.  The main work is going to be modifying https://github.com/mendersoftware/meta-mender-community/tree/kirkstone-next/meta-mender-tegra to move away from cboot references and to UEFI instead, especially the content in https://github.com/mendersoftware/meta-mender-community/tree/kirkstone-next/meta-mender-tegra/recipes-mender/tegra-state-scripts/files.  I'd probably start by getting the state scripts at a point where they can do what you need them to do on a Jetpack 5 build.  This may need some tweaks with Jetpack 5.1 when NVIDIA releases their working OTA A/B scheme.

**Reply by cmcquinn** [reply]
I'm getting everything set up to start working on this, but a `demo-image-base` built from the latest `kirkstone` branch of `tegra-demo-distro` does not seem to boot on my Xavier NX Devkit. I first tried it on an SD card module, and then rebuilt for an eMMC module, but neither seem to boot. The fan spins for several seconds after it finishes flashing, but ssh does not work and I don't see a serial port. Is this a known issue?

**Reply by madisongh** [reply]
I just flashed and booted `demo-image-base` on my `jetson-xavier-nx-devkit` and `jetson-xavier-nx-devkit-emmc`, from kirkstone latest.  I did find an issue with `dosdcard.sh` failing when trying to populate the SDcard for my dev kit, but the image booted up OK on both devices.

**Reply by cmcquinn** [reply]
It turns out that the images were booting, but I just didn't have any way of knowing that. I added the `l4t-usb-device-mode` package to the image, rebuilt, and reflashed, and the system comes up and I can ssh into it. I will start playing around with mender and see what I can figure out.

**Reply by cmcquinn** [reply]
I have been digging through the `tegra-state-scripts` and I have a few questions. `redundant-boot-install-script-uboot` saves u-boot variables before applying the bootloader update. I assume we need to do something similar with UEFI variables?

Reading through the [NVIDIA Developer Guide](https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/SD/Bootloader/UpdateAndRedundancy.html#update-bootloader), it shows a process for doing bootloader updates using `nv_update_engine`. Should we follow the same process for updating the bootloader with mender? The current state scripts use `tegra-bootloader-update` for updating the bootloader, but there aren't any references to `tegra-bootloader-update` in the `kirkstone` version of the `meta-tegra` layer. Is `nv_update_engine` a replacement for this?

**Reply by dwalkes** [reply]
Thanks @cmcquinn I'll take a stab at these and perhaps some folks who know more than I do can jump in with clarifications/corrections.

> redundant-boot-install-script-uboot saves u-boot variables before applying the bootloader update. I assume we need to do something similar with UEFI variables?

I think the save and restore [here](https://github.com/mendersoftware/meta-mender-community/blob/kirkstone-next/meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-install-script-uboot#L52-L57) is meant to handle cases where locations of uboot variables change between releases.  This shouldn't be relevant for the UEFI case, so I don't think it will be needed.

> Reading through the [NVIDIA Developer Guide](https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/SD/Bootloader/UpdateAndRedundancy.html#update-bootloader), it shows a process for doing bootloader updates using nv_update_engine. Should we follow the same process for updating the bootloader with mender?

Probably eventually, although I think the swupdate method Kurt described in https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2022-10-13 and https://gitter.im/OE4T/jetpack-5-test?at=6309262172ad51741fa1d436 isn't using this and is just updating specific partitions directly, then twiddling UEFI variables and the extlinux.conf file in the rootfs.

If you do use the `nv_update_engine` I think you'll need to make sure you update both slots at once due to the limitations in https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/SD/Bootloader/UpdateAndRedundancy.html#update-engine with r35.1, and then rework that with whatever NVIDIA comes up with in the next release for individual A/B updates.

> there aren't any references to tegra-bootloader-update in the kirkstone version of the meta-tegra layer. Is nv_update_engine a replacement for this?

There's no tegra-boot-loader-update for Jetpack 5 yet, and it's unclear if we will have/need one.  If we don't create a new one for Jetpack 5 and beyond we'll probably need to do our own hacks/workarounds for the [issues this tool fixed in Jetpack 4 and earlier](https://github.com/OE4T/tegra-boot-tools/blob/master/doc/tegra-bootloader-update.md).

**Reply by cmcquinn** [reply]
Do we need to keep the platform-specific `do_compile` functions in https://github.com/mendersoftware/meta-mender-community/blob/kirkstone/meta-mender-tegra/recipes-mender/tegra-state-scripts/tegra-state-scripts_1.0.bb?

**Reply by dwalkes** [reply]
You will ultimately need a state script copy step somewhere, however the first priority should be figuring out what content to put in the state script and getting everything working from the command line.

**Reply by cmcquinn** [reply]
I have an initial attempt at updating the state scripts at https://github.com/cmcquinn/meta-mender-community/tree/tegra-jetpack-5/meta-mender-tegra/recipes-mender/tegra-state-scripts-edk2. How do I test them?

**Reply by dwalkes** [reply]
You'd test them by running them as scripts, which is what Mender will ultimately do.  However I think you are referencing things which aren't available yet, for instance I don't think the [slot checks here will work yet](https://github.com/cmcquinn/meta-mender-community/blob/tegra-jetpack-5/meta-mender-tegra/recipes-mender/tegra-state-scripts-edk2/files/edk2_printenv#L5) and I don't think it's going to be possible to [only update one slot](https://github.com/cmcquinn/meta-mender-community/blob/tegra-jetpack-5/meta-mender-tegra/recipes-mender/tegra-state-scripts-edk2/files/redundant-boot-install-script#L28-L34) yet, I think you'll need to update both slots at once with either multiple calls to tegra-bootloader-update or using `nv_update_engine --forced-install` as described in https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/SD/Bootloader/UpdateAndRedundancy.html#update-bootloader.  I suggest getting the scripts tested and working individually first before you try the mender update though, they are difficult to debug once they are built into the update package.

**Reply by dwalkes** [reply]
Possibly related discussion in https://github.com/OE4T/meta-tegra/discussions/1115 - I'm wondering if we can skip the entire A/B integration headaches with NVIDIA going forward and use a similar recovery partition scheme for updates:
1) Across L4T releases/handling partition layout changes
2) Written to other storage devices like nvme
3) Optionally applying encryption
4) As a basis for image updates instead of requiring A/B partitioning.
